### DISCUSSION TITLE
Pgvecto.rs just blocked my plans to move Immich. How can I contribute to a more flexible model?

### DISCUSSION THREAD

**Discussion by btajuddin** [immich-app/immich#5954] [Q&A] [answered]
First off, I want to say how much I appreciate this app. I have really enjoyed using it and seeing it grow.

That said, I was planning to move Immich into AWS in the next couple weeks since I have some reserved instances that just opened up and are already paid for. Unfortunately, the recent change to require pgvecto.rs means that I cannot use my reserved postgres host.

I think it would be great if we could provide more flexibility around this with some options. For example, pgvector is a [supported extension in AWS](https://docs.aws.amazon.com/AmazonRDS/latest/PostgreSQLReleaseNotes/postgresql-extensions.html), so we perhaps we could support that via configuration flag?

I've also been tinkering in my spare time (what little I have) with decoupling the `IStorageRepository` from the `fs` types so that it could be swapped out for a different implementation (say...S3/minio/etc.).

My point is that I like this application and want to support it, so how can I help?

**Comment by alextran1502** [comment]
More information/context on why we are using pgvecto.rs https://github.com/immich-app/immich/discussions/5830

**Reply by btajuddin** [reply]
Yes, I saw that. It simply becomes a blocker for anyone who wants to run a more durable setup than postgres in a container. I'm offering to help, so let's try to find a way to help everyone.

**Comment by mertalev** [comment]
In discussing the migration to pgvecto.rs, we considered 1. supporting pgvector for environments where pgvecto.rs can't be used (or where it's difficult to set up), or 2. allowing users who disable ML to use Immich without either extension. We decided against (1) because we considered it a maintenance burden to maintain compatibility with both extensions. We did try (2), but ran into issues with database migrations that made it tricky to get right. I'm open to exploring (2) more since I do think it's a viable option.

**Reply by btajuddin** [reply]
Thanks for the options. I'd be willing to give option 2 a look if you have any pointers around what went wrong the last time you tried.

**Reply by mertalev** [reply, marked as answer]
Sure! To give a bit of context, we use TypeORM to connect to Postgres and run our migrations. Both the connection to Postgres and the migrations are configured [here](https://github.com/immich-app/immich/blob/b94b59e49d7532a2503e2aadc6a6dfc76725c531/server/src/infra/database.config.ts#L27) as a `DataSource`. This object is used in four ways: at startup, the server uses it to connect to Postgres and run any pending migrations. For E2E, it's used to initialize a test database. TypeORM also relies on this object for CLI commands, which we use for two GH Actions checks. Lastly, it's also used by TypeORM to synchronize database entities.

Our basic idea was to separate out the pgvecto.rs migrations into their own folder and make a separate `DataSource` for them. This would only be used to run migrations; all other queries would use the "main" `DataSource`. At startup, we would wait for the main one to initialize and query Postgres for whether pgvecto.rs is available. We would only run the vector migrations if it's available: if not, we let the server start up and allow any ML features that rely on pgvecto.rs to fail. [This](https://github.com/immich-app/immich/blob/af20c9adbb3c40543f3009be292ff04433078b8d/server/src/infra/database.config.ts) is what that more or less what that looked like.

This worked for the actual startup, but e2e and CLI didn't work properly with this setup. The e2e database didn't get properly initialized (see [here](https://github.com/immich-app/immich/actions/runs/6963546795/job/18949341045)), while for CLI, TypeORM couldn't reconcile the fact that `dataSource` was uninitialized at first and created asynchronously. This led to [this](https://github.com/immich-app/immich/actions/runs/6963546795/job/18949341833) error. The situation with entities also introduced a layer of ambiguity: depending on whether the vector migrations have run, the entities would have different columns; one entity may not even exist.

There are some changes since then that will help with implementing this.

For one thing, we now have a proper database [service](https://github.com/immich-app/immich/blob/83da1c74fce2776d3902a38dc80a11c733682245/server/src/domain/database/database.service.ts) and [repository](https://github.com/immich-app/immich/blob/83da1c74fce2776d3902a38dc80a11c733682245/server/src/infra/repositories/database.repository.ts#L7C24-L7C24) that handle similar startup checks. 
We use dependency injection for both normal startup and e2e and wait for these checks to complete before it can move on. This makes the e2e aspect of things much easier. 

The solution to the CLI would likely come down to making a separate file just for it with a data source that runs both the standard migrations and the vector ones. There should also be a check that runs without the vector migrations to make sure it still works.

The [SmartInfoEntity](https://github.com/immich-app/immich/blob/main/server/src/infra/entities/smart-info.entity.ts) and [SmartSearchEntity](https://github.com/immich-app/immich/blob/main/server/src/infra/entities/smart-search.entity.ts) entities have synchronization disabled, and the [AssetFaceEntity](https://github.com/immich-app/immich/blob/main/server/src/infra/entities/asset-face.entity.ts) entity no longer selects the embedding column. As long as we're careful about how we use the entities, it should be possible to avoid issues here.

I hope this gives you enough to work with. I'm also totally open to another approach if you have anything in mind.

**Reply by btajuddin** [reply]
Yeah, this is definitely getting into a world where ORMs start breaking down. This is definitely a complex problem to approach without creating parallel copies of large chunks of the source (which is obviously not an acceptable solution). I will keep looking at this, but it will definitely take some time to try to coax TypeORM to do what I'd like it to.

**Reply by mertalev** [reply]
Yeah, the biggest source of complexity is just making it work with TypeORM. One suggestion I have is to try to make the no-vectors DB layout as similar as possible to the vectors version.

In my earlier comment, I mentioned that "the entities would have different columns; one entity may not even exist", but this was because my attempt treated the vector migrations as all-or-nothing. Since these migrations included the move from the `smart_info` table to the new `smart_search` table, those changes would never happen if the vectors migrations didn't run.

But in reality, you could still apply that part of the migration to make sure the `smart_search` table exists and both `smart_info` and `smart_search` have the right columns. The vector migrations would just convert the embedding column's data type and add the index for it. That would at least help with keeping the entities in sync with the database.

**Reply by mertalev** [reply]
If you have an idea for a non-hacky way to implement it, I'm also open to pgvector support similar to your idea of decoupling `IStorageRepository`. I'm not sure if it'd be easier or harder than the no-vectors approach.

**Reply by btajuddin** [reply]
That's an idea that crossed my mind too. If we treated the vectors data in the DB as a separate "schema" from TypeORM's perspective, we might be able to cut it (and relevant interfaces) out cleanly with a config option. That would probably require some in-software joins that might hit performance a bit. We'd also have to figure out exactly how to mangle the migrations so that new and existing users could move forward without breaking one or the other.

**Reply by mertalev** [reply]
We already somewhat trick TypeORM. As far as TypeORM knows, the embedding columns are regular float arrays: everything on top of that is done with a combination of tailored queries and disabling synchronization. I think it'd be doable to separate the migrations so the tables and columns for each entity all exist, but the extension-specific changes can be in the repositories and handled at startup. You wouldn't need software joins, you'd just need to make sure each `ISmartInfoRepository` implementation has the logic needed to make these changes and keep extension-specific logic contained in that repository.

**Comment by btajuddin** [comment]
Thanks. I'll start looking through this in the next couple days.

**Comment by gaocegege** [comment]
Hi, I am one of the maintainers at github.com/tensorchord/pgvecto.rs. We're partnering with AWS to offer a pgvecto.rs-supported Postgres service on the AWS Marketplace. I wonder if this will help you.

**Reply by btajuddin** [reply]
Probably not. I'm assuming (and correct me if I'm wrong) that you're offering an AMI for postgres with pgvecto.rs. In my view, that's the same as running a container. What I would need is for pgvecto.rs to run in RDS postgres instances. I'm not aware of a way for a marketplace app to provide extensions into the runtime of RDS instances, but I'd be happy to be wrong about that.

**Reply by mertalev** [reply]
I think it needs to be added to [this](https://docs.aws.amazon.com/AmazonRDS/latest/PostgreSQLReleaseNotes/postgresql-extensions.html) list for RDS.

Based on [this](https://aws.amazon.com/rds/postgresql/faqs/), AWS suggests emailing `rds-postgres-extensions-request@amazon.com` to request an extension to be added. Given the focus on GenAI and how much attention pgvector has gotten recently because of it, I wouldn't be surprised if they were receptive to adding an extension that has so many advantages over it.

Edit: Oops, fixed the email address.

**Reply by gaocegege** [reply]
@mertalev Hi there! Thanks for sharing this information, I will give it a try.

**Reply by gaocegege** [reply]
Besides this, we will ensure that pgvecto.rs remains compatible with pgvector in terms of basic features, in the near future, while allowing for potential differences or enhancements in advanced functionalities.

For example, you could run query statements as pgvector does. But you will need other statements to use features like vbase and other advanced features.

**Comment by SISheogorath** [comment]
Just want to join the chorus: Talking with sysadmins around me about this, we always end up with "Yeah, you don't want some custom 0.x extension in your postgresql instance".

I mean, dropping facial recognition would take a big chunk of usefulness out of immich, but at this point I rather have no facial recognition for a while (maybe until the pgvector.rs extension goes mainstream and becomes stable?) rather than sticking around on an old version of immich with the pending doom of apps going out of compatibility.

There are various thoughts in the room. From just ditching immich in favour of Photoprism or alike to forking v1.90.2 and built a compatible API using the old typesense stack. Databases are a solved problem in most stacks… until someone comes around and has to do them differently.

I understand that from a developer side of things, it's an easy choice, it probably makes a lot of things easier to implement and benefits the whole software greatly. It's free software, you are free to do what you like and you don't owe anyone anything. Still thanks for reading these concerns.

**Reply by mertalev** [reply]
In fairness, pgvector is also a 0.* extension and widely supported and deployed. In fact, I would argue that it's more stable than Typesense and many other experimental vector databases. A search in this repo should be enough to demonstrate the litany of stability and data integrity issues surrounding Typesense. The only reason data loss hasn't been an issue there is because we always stored everything in Postgres too.

**Reply by SISheogorath** [reply]
Well, I've done it. But let's be honest, it's not really beautiful…

https://git.shivering-isles.com/shivering-isles/infrastructure-gitops/-/compare/176fa49766644f921660e7d553dc04c5ee08944e...c7e6b1c7b719ac3cadd3853b0cfd378edd9676aa?from_project_id=233&straight=false

**Comment by Caian** [comment]
As someone that uses immich without ML or typesense, I think that forcing users into a beta database plugin for an optional feature was a poor decision.

**Comment by doctorpangloss** [comment]
can this be gracefully disabled or removed? or can you explicitly document that you do not support AWS Aurora RDS? we took a lot of brain damage trying to deploy this, only to find this out

**Reply by mertalev** [reply]
We haven't documented this yet, but you can add the `DB_VECTOR_EXTENSION=pgvector` environmental variable to the server and microservices containers to use pgvector instead.

**Reply by dani** [reply]
But this is not feasible for an existing install, already running with pgvecto.rs AFAIK. I'd love to have a migration script to move to pgvector (but loosing all the face detection would be a pain)

**Reply by gaocegege** [reply]
We have a managed service for pgvecto.rs. https://pgvecto.rs/ You could take a look if you are interested.

**Reply by dani** [reply]
Thanks, but the whole point of using immich, is to not rely on external managed service

**Comment by mmomjian** [comment]
This is now supported.
